src/control/server/mgmt_client.go

```diff
@@ -94,43 +91,89 @@ func (msc *mgmtSvcClient) LeaderAddress() (string, error) {
 }
 
 func (msc *mgmtSvcClient) Join(ctx context.Context, req *mgmtpb.JoinReq) (resp *mgmtpb.JoinResp, joinErr error) {
-	joinErr = msc.withConnection(ctx, func(ctx context.Context, ap string, pbClient mgmtpb.MgmtSvcClient) error {
-		prefix := fmt.Sprintf("join(%s, %+v)", ap, *req)
-		msc.log.Debugf(prefix + " begin")
-		defer msc.log.Debugf(prefix + " end")
-
-		if req.Addr == "" {
-			req.Addr = msc.cfg.ControlAddr.String()
-		}
-
-		for {
-			select {
-			case <-ctx.Done():
-				return errors.Wrap(ctx.Err(), prefix)
-			default:
+	ap, err := msc.LeaderAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	joinErr = msc.withConnection(ctx, ap,
+		func(ctx context.Context, pbClient mgmtpb.MgmtSvcClient) error {
+
+			prefix := fmt.Sprintf("join(%s, %+v)", ap, *req)
+			msc.log.Debugf(prefix + " begin")
+			defer msc.log.Debugf(prefix + " end")
+
+			if req.Addr == "" {
+				req.Addr = msc.cfg.ControlAddr.String()
 			}
 
-			resp, err := pbClient.Join(ctx, req)
-			if err != nil {
-				msc.log.Debugf(prefix+": %v", err)
-			} else {
-				// TODO: Stop retrying upon certain errors (e.g., "not
-				// MS", "rank unavailable", and "excluded").
-				if resp.Status != 0 {
-					msc.log.Debugf(prefix+": %d", resp.Status)
+			for {
```

Comment:
I think it's fine to land this as-is for now to continue forward progress, but the fact that these two function bodies are nearly identical seems to indicate an opportunity to further DRY things up.

Reply:
agreed